### ISSUE TITLE
Integer HLSL dot() should not use SPIR-V OpDot

### ISSUE THREAD

**Issue by lostgoat** [KhronosGroup/glslang#1648]
We seem to have a behaviour difference in how the output of a dot product is treated between fxc and glslangValidator.

In fxc the output of the dot product matches the type of the resulting variable, in glslangValidator it matches the type of the inputs.

Here is an example shader that produces spirv with validation warnings:

```
float4 VS_MAIN(float4 i : POSITION) : POSITION
{
	float4 ret = float4( 0 , 1, 2, 3 );

	int2 vVector = int2( 1, 2 );

	// This produces the following warning
	// error: line 62: Expected float scalar type as Result Type: Dot
	//    %28 = OpDot %int %26 %27
	ret.r = dot( vVector.xy, vVector.xy );

	ret.g = dot( float2( vVector.xy ), float2( vVector.xy ) );

    return ret;
}
```

I'm compiling it with:
```
glslangValidator -D -S vert -V -e VS_MAIN -o out.spv glslang_fail.hlsl && spirv-val out.spv
```

**Comment by lostgoat** [comment]
@danginsburg fyi

**Comment by johnkslang** [comment]
> In fxc the output of the dot product matches the type of the resulting variable, in glslangValidator it matches the type of the inputs.

When dot is computed by either compiler, the only type information should be the inputs. Perhaps there is a differing implicit conversion across the `=`, which is a subsequent step to computing `dot`.

**Comment by johnkslang** [comment]
Could either 

1. convert the args to float, or
2. emulate dot with integer multiply and integer add

FXC emulates dot for integer inputs with integer math, and I recommend that glslang can do the same.

**Comment by lostgoat** [comment]
@johnkslang Thanks for breaking down the issue into better words than I could : )

**Comment by johnkslang** [comment]
It looks like this is correctly handled in #1497. I'm reviewing that now, and hopefully it will be in shortly.

**Comment by TiemoJung** [comment]
was fixed with #1583 why #1497 marks this as new is beyond me, a git thing or something
